src/mapping/query/functions.rs

```diff
@@ -163,12 +164,55 @@ impl Function for ToBooleanFn {
 
 #[derive(Debug)]
 pub(in crate::mapping) struct ToTimestampFn {
-    conversion: Conversion,
     query: Box<dyn Function>,
     default: Option<Value>,
 }
 
 impl ToTimestampFn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, default: Option<Value>) -> Self {
+        Self { query, default }
+    }
+}
+
+impl Function for ToTimestampFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        use chrono::{DateTime, TimeZone, Utc};
+
+        let value_to_timestamp = |v| {
+            match v {
+                Value::Bytes(b) => DateTime::parse_from_rfc3339(&String::from_utf8_lossy(&b))
+                    .map(|dt| dt.with_timezone(&Utc))
+                    .map_err(|_| "cannot parse string as RFC3339".to_owned()),
+                Value::Integer(i) => Ok(Utc.timestamp(i, 0)),
+                Value::Timestamp(t) => Ok(t),
+                _ => Err("unable to parse non-string or integer type to timestamp".to_string()),
+            }
+            .map(Value::Timestamp)
+        };
```

Comment:
This closure does not appear to reference any of the context. If so, I think it should at least be converted to a plain function and possibly moved out of `execute`.